### ISSUE TITLE
`object*` is not iterable with `itertools.chain`

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#5172]
**Describe the bug**
Given the following code:

```python
from __future__ import annotations

from itertools import chain

times = [
    f'{12 if hour == 0 else hour!s}:{minute!s:0>2} {meridian}'
    for hour, minute, meridian in chain.from_iterable(
        chain.from_iterable(
            [(hour, minute, meridian) for minute in range(0, 60, 15)]
            for hour in range(12)
        )
        for meridian in ('am', 'pm')
    )
]
```

The following errors are reported:

```
foo.py
  foo.py:7:35 - error: "object*" is not iterable
    "__iter__" method not defined (reportGeneralTypeIssues)
  foo.py:7:9 - error: Type of "hour" is unknown (reportUnknownVariableType)
  foo.py:7:15 - error: Type of "minute" is unknown (reportUnknownVariableType)
  foo.py:7:23 - error: Type of "meridian" is unknown
```

**Expected behavior**
No errors should be reported

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line

Which version?
1.1.310

**Additional context**
This is probably related to #5126 as it started happening with version 1.308, but is not fixed in the latest version.

**Comment by erictraut** [comment]
Thanks for the bug report. Looks to be the same issue as https://github.com/microsoft/pyright/issues/5165. I suspect it's because the same generic function (in your code sample, `chain.from_iterable`) is called multiple times in a nested manner. I'll investigate further.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.312, which I just published. It will also be included in the next insiders release of pylance.

**Comment by bryanforbes** [comment]
@erictraut this one is happening again in 1.1.336

**Comment by erictraut** [comment]
@bryanforbes, it's unlikely to be the same issue. I added extensive regression tests for the previous issue at the time I fixed it. If you're seeing something similar, please open a new bug with repro steps.

**Comment by bryanforbes** [comment]
The only reason I asked here is because the same code above triggers the last three errors. I can open a new bug if you still would like one.
